### PR TITLE
BaseDenseAttention now supports attention dropout

### PR DESCRIPTION
**Pull request by claverru**
Added optional dropout applied to attention weights (scores)

### CODE REVIEW

**Comment by claverru** [comment]
> @googlebot I signed it!

**Comment by tanzhenyu** [comment]
I'm not sure I'm the best person to review this.
Can you find the author of dense attention?
https://github.com/tensorflow/community/blob/master/rfcs/20190115-dense-attention.md

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by claverru** [comment]
@roumposg 

https://github.com/tensorflow/models/blob/master/official/transformer/v2/attention_layer.py

I know its not a paper, but the official TensorFlow implementation. I was just morroring it.

And the results are slightly different from the "more conventional" dropout, given that it's applied to the attention weights. Think of it as the recurrent dropout the RNNs have.

On the other hand, I was about to contribute the implementation of the Relative Encoding (https://arxiv.org/abs/1803.02155, and with a better explanation https://arxiv.org/abs/1809.04281). Would it be a good idea?

**Review by roumposg — CHANGES_REQUESTED** [review]
Let's add a test. Because `dropout` is probabilistic, we can only test the shape of the output tensor. See the `test_shape` tests in https://github.com/tensorflow/tensorflow/blob/51c581749570584d504d9a748f1b6202d6adcced/tensorflow/python/keras/layers/dense_attention_test.py

**Comment by claverru** [comment]
@roumposg Thank you, I appreciate it. I will fix them as soon as I can. Also thanks to @tanzhenyu .

File: tensorflow/python/keras/layers/dense_attention.py [hunk 1]
```diff
@@ -46,6 +46,7 @@ class BaseDenseAttention(Layer):
     causal: Boolean. Set to `True` for decoder self-attention. Adds a mask such
       that position `i` cannot attend to positions `j > i`. This prevents the
       flow of information from the future towards the past.
+    dropout: Float between 0 and 1. Fraction of the units to drop for the attention scores.
```
[the remark this answers is not in the document]
Comment (claverru) [reply tensorflow/python/keras/layers/dense_attention.py]: Done

File: tensorflow/python/keras/layers/dense_attention.py [hunk 2]
```diff
@@ -59,6 +60,7 @@ class BaseDenseAttention(Layer):
       * query_mask: A boolean mask `Tensor` of shape `[batch_size, Tq]`.
         If given, the output will be zero at the positions where
         `mask==False`.
+      * training: Whether the layer is training or not. Used to apply dropout.
```
[the remark this answers is not in the document]
Comment (claverru) [reply tensorflow/python/keras/layers/dense_attention.py]: Done

File: tensorflow/python/keras/layers/dense_attention.py [hunk 3]
```diff
@@ -84,7 +87,7 @@ def _calculate_scores(self, query, key):
     """
     return NotImplementedError
 
-  def _apply_scores(self, scores, value, scores_mask=None):
+  def _apply_scores(self, scores, value, training, scores_mask=None):
```
[the remark this answers is not in the document]
Comment (claverru) [reply tensorflow/python/keras/layers/dense_attention.py]: Done

File: tensorflow/python/keras/layers/dense_attention.py [hunk 4]
```diff
@@ -112,10 +115,12 @@ def _apply_scores(self, scores, value, scores_mask=None):
       # Bias so padding positions do not contribute to attention distribution.
       scores -= 1.e9 * math_ops.cast(padding_mask, dtype=K.floatx())
     attention_distribution = nn.softmax(scores)
+    if training is True:
+	    weights = tf.nn.dropout(weights, rate=self.dropout)
```
[the remark this answers is not in the document]
Comment (claverru) [reply tensorflow/python/keras/layers/dense_attention.py]: Done

File: tensorflow/python/keras/layers/dense_attention.py [hunk 5]
```diff
@@ -112,10 +115,12 @@ def _apply_scores(self, scores, value, scores_mask=None):
...
     return math_ops.matmul(attention_distribution, value)
 
   # TODO(b/125916026): Consider exposing a __call__ method with named args.
-  def call(self, inputs, mask=None):
+  def call(self, inputs, training, mask=None):
```
[the remark this answers is not in the document]
Comment (claverru) [reply tensorflow/python/keras/layers/dense_attention.py]: Not sure since dropout is suposed to work differently in training mode and inference mode.

File: tensorflow/python/keras/layers/dense_attention.py [hunk 6]
```diff
@@ -139,7 +144,7 @@ def call(self, inputs, mask=None):
     else:
       causal_mask = None
     scores_mask = _merge_masks(v_mask, causal_mask)
-    result = self._apply_scores(scores=scores, value=v, scores_mask=scores_mask)
+    result = self._apply_scores(scores=scores, value=v, training, scores_mask=scores_mask)
```
[the remark this answers is not in the document]
Comment (claverru) [reply tensorflow/python/keras/layers/dense_attention.py]: Done

**Review by roumposg — CHANGES_REQUESTED** [review]
Thanks for working on this. Please, also add a test. Because `dropout` is probabilistic, we can only test the shape of the output tensor. See the `test_shape` tests in https://github.com/tensorflow/tensorflow/blob/51c581749570584d504d9a748f1b6202d6adcced/tensorflow/python/keras/layers/dense_attention_test.py and add a similar one, thanks.

Also, please add validation as I replied in one of the comments, thanks.

**Comment by claverru** [comment]
@roumposg I think it should be finished now, let me know please.

**Review by roumposg — DISMISSED** [review]

**Review by tanzhenyu — DISMISSED** [review]

**Review by tanzhenyu — APPROVED** [review]

**Comment by claverru** [comment]
@gbaned Is everything alright?

**Comment by gbaned** [comment]
@claverru Sorry for the late reply. It is pending with internal merging. Thanks!

**Comment by arthurflor23** [comment]
Any news?

**Comment by claverru** [comment]
> Any news?

@gbaned ?

**Comment by gbaned** [comment]
@arthurflor23, @claverru  This is failing some internal tests, please stay tuned. Thanks!

**Comment by tanzhenyu** [comment]
It broke internal tests so @roumposg submitted a separate commit [here](https://github.com/tensorflow/tensorflow/commit/2c0ab25cad67841e28218f3612801f9a2d673a0a)

**Comment by claverru** [comment]
Thank you @tanzhenyu , @roumposg !

**Comment by rthadur** [comment]
Seems auto-merge is not happening but the changes are now committed so we can close this. Thank you for the PR.
